File: android/app/src/main/java/com/modulestutorial/PackageTrackerModule.kt
package com.modulestutorial

import android.os.Handler
import android.os.Looper
import com.facebook.react.bridge.*
import java.util.*

data class Package(val id: String, val verificationCode: String, val distance: Int, val status: String)

val packages = listOf(Package("WS1", "WSROCKS", 5000, "moving"), Package("WS2", "WSRN", 0, "delivered"))

class PackageTrackerModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    @ReactMethod
    fun track(data: ReadableMap, promise: Promise) {
        val id = data.getString("id")
        val verificationCode = data.getString("verificationCode")

        if(id == null || verificationCode == null) {
            promise.reject("PACKAGE_NOT_FOUND", "Id and Verification code didn't match a package")

            return
        }

        val handler = Handler(Looper.getMainLooper())

        handler.postDelayed({
            val matchingPackage = packages.find { it.id == id && it.verificationCode == verificationCode }

            if (matchingPackage == null) {
                promise.reject("PACKAGE_NOT_FOUND", "Id and Verification code didn't match a package")

                return@postDelayed
            }

            val response = Arguments.createMap().apply {
                putMap("request", data)
                putInt("distance", matchingPackage.distance)
                putString("status", matchingPackage.status)
            }


            promise.resolve(response)

        }, 5000)
    }

    override fun getName(): String {
        return "PackageTrackerModule"
    }
}

File: ios/PackageTrackerModule.swift
//
//  PackageTrackerModule.swift
//  ModulesTutorial
//
//  Created by Jefferson Tavares de Pádua on 28/10/22.
//

import Foundation
import React

struct Package {
  let id: String
  let verificationCode: String
  let distance: Int
  let status: String
}

let packages: [Package] = [
  Package(id: "WS1", verificationCode:"WSROCKS", distance: 5000, status: "moving"),
  Package(id: "WS2", verificationCode:"WSRN", distance: 0, status: "delivered"),
]

@objc(PackageTrackerModule)
class PackageTrackerModule : NSObject {
  
  
  @objc
  func requiresMainQueueSetup() -> Bool {
      return true
  }
  
  @objc
  func track(_ data: NSDictionary,  resolver resolve: @escaping RCTPromiseResolveBlock,  rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard let id = data["id"] as? String, let verificationCode = data["verificationCode"] as? String else {
      reject("PACKAGE_NOT_FOUND", "Id and Verification code didn't match a package", nil)
      
      return
    }
    
    reject()
    
    let matchingPackage = packages.first{ $0.id == id && $0.verificationCode == verificationCode }
    
    DispatchQueue.main.asyncAfter(deadline: .now() + 5) {
      if let matched = matchingPackage {
        resolve(["request": data, "distance": matched.distance, "status": matched.status])
      } else {
        reject("PACKAGE_NOT_FOUND", "Id and Verification code didn't match a package", nil)
      }
    }
  }
}
